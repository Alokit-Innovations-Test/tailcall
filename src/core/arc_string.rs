use std::sync::Arc;

use headers::HeaderValue;
use schemars::gen::SchemaGenerator;
use schemars::schema::Schema;
use serde::{Deserialize, Deserializer, Serialize, Serializer};

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct ArcString(Arc<str>);

impl<'de> Deserialize<'de> for ArcString {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        String::deserialize(deserializer).map(ArcString::from)
    }
}

impl Serialize for ArcString {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        self.as_str().serialize(serializer)
    }
}

impl Default for ArcString {
    fn default() -> Self {
        ArcString("".into())
    }
}

impl AsRef<str> for ArcString {
    fn as_ref(&self) -> &str {
        &self.0
    }
}

impl From<&str> for ArcString {
    fn from(s: &str) -> Self {
        ArcString(s.into())
    }
}

impl From<&String> for ArcString {
    fn from(s: &String) -> Self {
        ArcString(s.clone().into())
    }
}

impl From<String> for ArcString {
    fn from(s: String) -> Self {
        ArcString(s.into())
    }
}

impl From<ArcString> for Vec<u8> {
    fn from(s: ArcString) -> Self {
        s.as_str().as_bytes().to_vec()
    }
}

impl TryFrom<ArcString> for HeaderValue {
    type Error = <HeaderValue as TryFrom<String>>::Error;

    fn try_from(s: ArcString) -> Result<Self, Self::Error> {
        s.as_str().parse()
    }
}

impl PartialEq<&str> for ArcString {
    fn eq(&self, other: &&str) -> bool {
        self.as_ref() == *other
    }
}

impl schemars::JsonSchema for ArcString {
    fn schema_name() -> String {
        <String as schemars::JsonSchema>::schema_name()
    }

    fn json_schema(gen: &mut SchemaGenerator) -> Schema {
        <String as schemars::JsonSchema>::json_schema(gen)
    }
}

impl std::fmt::Display for ArcString {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        self.as_str().fmt(f)
    }
}

impl ArcString {
    pub fn as_str(&self) -> &str {
        self.as_ref()
    }
}
