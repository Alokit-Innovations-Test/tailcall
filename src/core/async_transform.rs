use std::future::Future;

use super::valid::{Valid, Validator};

/// A configuration transformer that allows us to perform various
/// transformations on the configuration before it's further processed for
/// blueprint creation.
pub trait AsyncTransform {
    type Value: Send;
    type Error;

    fn transform(
        &self,
        value: Self::Value,
    ) -> impl Future<Output = Valid<Self::Value, Self::Error>> + Send;
}

/// A suite of common operators that are available for all transformers.
pub trait AsyncTransformerOps: AsyncTransform + Send {
    fn pipe<Other: AsyncTransform<Value = Self::Value, Error = Self::Error>>(
        self,
        other: Other,
    ) -> Pipe<Self, Other>
    where
        Self: Sized,
    {
        Pipe(self, other)
    }

    fn generate(&self) -> impl Future<Output = Valid<Self::Value, Self::Error>> + Send
    where
        Self: Send + Sync,
        Self::Value: std::default::Default,
    {
        async move { self.transform(Self::Value::default()).await }
    }

    fn when(self, cond: bool) -> When<Self>
    where
        Self: Sized,
    {
        When(self, cond)
    }
}

impl<T: AsyncTransform + Sync + Send> AsyncTransformerOps for T {}

pub struct When<A>(A, bool);
impl<A: AsyncTransform + Sync + Send> AsyncTransform for When<A> {
    type Value = A::Value;
    type Error = A::Error;

    async fn transform(
        &self,
        value: Self::Value,
    ) -> Valid<Self::Value, Self::Error> {
        if self.1 {
            self.0.transform(value).await
        } else {
            Valid::succeed(value)
        }
    }
}

/// Represents a composition of two transformers.
pub struct Pipe<A, B>(A, B);

impl<A, E, X, Y> AsyncTransform for Pipe<X, Y>
where
    X: AsyncTransform<Value = A, Error = E> + Send + Sync,
    Y: AsyncTransform<Value = A, Error = E> + Send + Sync,
    A: Sync + Send,
    E: Sync + Send,
{
    type Value = A;
    type Error = E;

    async fn transform(
        &self,
        value: Self::Value,
    ) -> Valid<Self::Value, Self::Error> {
        let result = self.0.transform(value).await;
        match result.to_result() {
            Ok(result) => self.1.transform(result).await,
            Err(err) => Valid::from_validation_err(err),
        }
    }
}

/// Represents an empty transformer.
pub struct Default<A, E>(std::marker::PhantomData<(A, E)>);

impl<A: Send + Sync, E: Send + Sync> AsyncTransform for Default<A, E> {
    type Value = A;
    type Error = E;
    async fn transform(
        &self,
        value: Self::Value,
    ) -> Valid<Self::Value, Self::Error> { Valid::succeed(value) }
}

// pub fn default<A, E>() -> Default<A, E> {
//     Default(std::marker::PhantomData)
// }
