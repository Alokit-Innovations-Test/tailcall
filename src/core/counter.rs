use std::cell::RefCell;

#[allow(unused)]
#[derive(Default)]
pub struct Counter(RefCell<usize>);
impl Counter {
    pub fn next(&self) -> usize {
        self.0.replace_with(|a| *a + 1)
    }
}
