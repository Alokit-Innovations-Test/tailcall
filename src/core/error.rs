use std::fmt::Display;
use std::string::FromUtf8Error;

use derive_more::From;
use inquire::InquireError;

use super::config::UnsupportedConfigFormat;

#[derive(From, thiserror::Error, Debug)]
pub enum Error {
    #[error("Worker Error: {}", _0)]
    Worker(worker::Error),

    #[error("File Error: {}", _0)]
    File(file::Error),

    #[error("Inquire Error: {}", _0)]
    Inquire(InquireError),

    #[error("Serde Yaml Error: {}", _0)]
    SerdeYaml(serde_yaml::Error),

    #[error("Package name is required")]
    PackageNameNotFound,

    #[error("Protox Parse Error")]
    ProtoxParse(protox_parse::ParseError),

    #[error("Unable to extract content of google well-known proto file")]
    GoogleProtoFileContentNotExtracted,

    #[error("Utf8 Error: {}", _0)]
    Utf8(FromUtf8Error),

    #[error("Unsupported Config Format")]
    UnsupportedConfigFormat(UnsupportedConfigFormat),

    #[error("Unsupported File Format")]
    UnsupportedFileFormat,

    #[error("Serde Json Error: {}", _0)]
    SerdeJson(serde_json::Error),

    #[error("Unable to determine path")]
    PathDeterminationFailed,

    #[error("Schema mismatch Error")]
    SchemaMismatch,

    #[error("Error: {}", _0)]
    Anyhow(anyhow::Error),
}

pub mod worker {
    use std::sync::Arc;

    use derive_more::{DebugCustom, From};
    use tokio::task::JoinError;

    #[derive(From, DebugCustom, Clone)]
    pub enum Error {
        #[debug(fmt = "Failed to initialize worker")]
        InitializationFailed,

        #[debug(fmt = "Worker communication error")]
        Communication,

        #[debug(fmt = "Serde Json Error: {}", _0)]
        SerdeJson(Arc<serde_json::Error>),

        #[debug(fmt = "Request Clone Failed")]
        RequestCloneFailed,

        #[debug(fmt = "Hyper Header To Str Error: {}", _0)]
        HyperHeaderStr(Arc<hyper::header::ToStrError>),

        #[debug(fmt = "JS Runtime Stopped Error")]
        JsRuntimeStopped,

        #[debug(fmt = "CLI Error : {}", _0)]
        CLI(String),

        #[debug(fmt = "Join Error : {}", _0)]
        Join(Arc<JoinError>),

        #[debug(fmt = "Runtime not initialized")]
        RuntimeNotInitialized,

        #[debug(fmt = "{} is not a function", _0)]
        #[from(ignore)]
        InvalidFunction(String),

        #[debug(fmt = "Rquickjs Error: {}", _0)]
        #[from(ignore)]
        Rquickjs(String),

        #[debug(fmt = "Deserialize Failed: {}", _0)]
        #[from(ignore)]
        DeserializeFailed(String),

        #[debug(fmt = "globalThis not initialized: {}", _0)]
        #[from(ignore)]
        GlobalThisNotInitialised(String),

        #[debug(
            fmt = "Error: {}\nUnable to parse value from js function: {} maybe because it's not returning a string?",
            _0,
            _1
        )]
        FunctionValueParseError(String, String),

        #[debug(fmt = "Error : {}", _0)]
        Anyhow(Arc<anyhow::Error>),
    }

    impl From<serde_json::Error> for Error {
        fn from(error: serde_json::Error) -> Self {
            Error::SerdeJson(Arc::new(error))
        }
    }

    impl From<hyper::header::ToStrError> for Error {
        fn from(error: hyper::header::ToStrError) -> Self {
            Error::HyperHeaderStr(Arc::new(error))
        }
    }

    impl From<JoinError> for Error {
        fn from(error: JoinError) -> Self {
            Error::Join(Arc::new(error))
        }
    }

    impl From<anyhow::Error> for Error {
        fn from(error: anyhow::Error) -> Self {
            Error::Anyhow(Arc::new(error))
        }
    }

    pub type Result<A> = std::result::Result<A, Error>;
}

impl Display for worker::Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            worker::Error::InitializationFailed => write!(f, "Failed to initialize worker"),
            worker::Error::Communication => write!(f, "Worker communication error"),
            worker::Error::SerdeJson(error) => write!(f, "Serde Json Error: {}", error),
            worker::Error::RequestCloneFailed => write!(f, "Request Clone Failed"),
            worker::Error::HyperHeaderStr(error) => {
                write!(f, "Hyper Header To Str Error: {}", error)
            }
            worker::Error::JsRuntimeStopped => write!(f, "JS Runtime Stopped Error"),
            worker::Error::CLI(msg) => write!(f, "CLI Error: {}", msg),
            worker::Error::Join(error) => write!(f, "Join Error: {}", error),
            worker::Error::RuntimeNotInitialized => write!(f, "Runtime not initialized"),
            worker::Error::InvalidFunction(function_name) => {
                write!(f, "{} is not a function", function_name)
            }
            worker::Error::Rquickjs(error) => write!(f, "Rquickjs error: {}", error),
            worker::Error::DeserializeFailed(error) => write!(f, "Deserialize Failed: {}", error),
            worker::Error::GlobalThisNotInitialised(error) => write!(f, "globalThis not initialized: {}", error),
            worker::Error::FunctionValueParseError(error, name) => write!(f, "Error: {}\nUnable to parse value from js function: {} maybe because it's not returning a string?", error, name),
            worker::Error::Anyhow(msg) => write!(f, "Error: {}", msg),
        }
    }
}

pub mod file {
    use std::string::FromUtf8Error;

    use derive_more::{DebugCustom, From};

    #[derive(From, DebugCustom)]
    pub enum Error {
        #[debug(fmt = "No such file or directory (os error 2)")]
        NotFound,

        #[debug(fmt = "No permission to access the file")]
        NoPermission,

        #[debug(fmt = "Access denied")]
        AccessDenied,

        #[debug(fmt = "Invalid file format")]
        InvalidFormat,

        #[debug(fmt = "Invalid file path")]
        InvalidFilePath,

        #[debug(fmt = "Invalid OS string")]
        InvalidOsString,

        #[debug(fmt = "Failed to read file : {}", _0)]
        FileReadFailed(String),

        #[debug(fmt = "Failed to write file : {}", _0)]
        #[from(ignore)]
        FileWriteFailed(String),

        #[debug(fmt = "Std IO Error: {}", _0)]
        StdIO(std::io::Error),

        #[debug(fmt = "Utf8 Error: {}", _0)]
        Utf8(FromUtf8Error),

        #[debug(fmt = "File writing not supported on Lambda.")]
        LambdaFileWriteNotSupported,

        #[debug(fmt = "Cannot write to a file in an execution spec")]
        ExecutionSpecFileWriteFailed,

        #[debug(fmt = "Cloudflare Worker Execution Error : {}", _0)]
        #[from(ignore)]
        Cloudflare(String),

        #[debug(fmt = "Error : {}", _0)]
        Anyhow(anyhow::Error),
    }

    pub type Result<A> = std::result::Result<A, Error>;
}

impl Display for file::Error {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            file::Error::NotFound => write!(f, "No such file or directory (os error 2)"),
            file::Error::NoPermission => write!(f, "No permission to access the file"),
            file::Error::AccessDenied => write!(f, "Access denied"),
            file::Error::InvalidFormat => write!(f, "Invalid file format"),
            file::Error::InvalidFilePath => write!(f, "Invalid file path"),
            file::Error::InvalidOsString => write!(f, "Invalid OS string"),
            file::Error::FileReadFailed(path) => write!(f, "Failed to read file: {}", path),
            file::Error::FileWriteFailed(path) => write!(f, "Failed to write file: {}", path),
            file::Error::StdIO(error) => write!(f, "Std IO Error: {}", error),
            file::Error::Utf8(error) => write!(f, "Utf8 Error: {}", error),
            file::Error::LambdaFileWriteNotSupported => {
                write!(f, "File writing not supported on Lambda.")
            }
            file::Error::ExecutionSpecFileWriteFailed => {
                write!(f, "Cannot write to a file in an execution spec")
            }
            file::Error::Cloudflare(error) => {
                write!(f, "Cloudflare Worker Execution Error: {}", error)
            }
            file::Error::Anyhow(msg) => write!(f, "Error: {}", msg),
        }
    }
}

pub type Result<A, E> = std::result::Result<A, E>;
