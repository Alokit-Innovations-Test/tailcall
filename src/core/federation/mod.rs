pub mod subgraph_reader;
