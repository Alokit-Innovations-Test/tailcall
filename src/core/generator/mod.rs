mod from_json;
pub mod config;
mod from_proto;
mod generator;
mod graphql_type;
mod json;
mod proto;
mod from_generalized_config;
pub mod source;

pub use from_json::{from_json, ConfigGenerationRequest};
pub use json::NameGenerator;
pub use generator::Generator;
pub use from_generalized_config::FromGeneralizedConfig;
