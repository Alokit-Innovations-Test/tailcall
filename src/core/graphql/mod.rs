mod data_loader;
mod request_template;
mod error;

pub use data_loader::*;
pub use request_template::*;
