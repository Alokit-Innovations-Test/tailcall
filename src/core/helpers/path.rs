use std::fs;
use std::path::Path;

use pathdiff::diff_paths;

/// Checks if file or folder already exists or not.
pub fn is_exists(path: &str) -> bool {
    fs::metadata(path).is_ok()
}

/// Expects both paths to be absolute and returns a relative path from `from` to
/// `to`. expects `from`` to be directory.
pub fn to_relative_path(from: &Path, to: &str) -> Option<String> {
    let from_path = Path::new(from).to_path_buf();
    let to_path = Path::new(to).to_path_buf();

    // Calculate the relative path from `from_path` to `to_path`
    diff_paths(to_path, from_path).map(|p| p.to_string_lossy().to_string())
}
