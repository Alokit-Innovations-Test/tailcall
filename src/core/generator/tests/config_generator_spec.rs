use std::path::Path;

use serde::{Deserialize, Serialize};
use tailcall::core::{
    blueprint::Blueprint,
    config,
    generator::{
        source::ImportSource, GeneratorConfig, GeneratorInput, GeneratorReImpl, InputSource,
        Resolved, UnResolved,
    },
    proto_reader::ProtoReader,
    resource_reader::ResourceReader,
};
use tokio::runtime::Runtime;

datatest_stable::harness!(
    run_config_generator_spec,
    "tailcall-fixtures/fixtures/generator/",
    r"^.*\.json"
);

pub fn run_config_generator_spec(path: &Path) -> datatest_stable::Result<()> {
    let path = path.to_path_buf();
    let runtime = Runtime::new().unwrap();
    runtime.block_on(async move {
        run_test(&path.to_string_lossy()).await?;
        Ok(())
    })
}
#[derive(Serialize, Deserialize)]
struct GeneratorTest {
    config: GeneratorConfig<UnResolved>,
    #[serde(default)]
    resolvers: serde_json::Value,
}

/// performs all the i/o's required in the config file and generates
/// concrete vec containing data for generator.
async fn resolve_io(
    config: GeneratorConfig<Resolved>,
    resolvers: serde_json::Value,
) -> anyhow::Result<Vec<GeneratorInput>> {
    let mut generator_type_inputs = vec![];
    let runtime = tailcall::cli::runtime::init(&Blueprint::default());

    let reader = ResourceReader::cached(runtime.clone());
    let proto_reader = ProtoReader::init(reader.clone(), runtime.clone());

    for input in config.input.iter() {
        match &input.source {
            InputSource::Import { src, _marker } => {
                let source = ImportSource::detect(&src)?;
                match source {
                    ImportSource::Url => {
                        let content = resolvers
                            .get(src)
                            .unwrap_or(&serde_json::Value::Null)
                            .to_owned();
                        generator_type_inputs
                            .push(GeneratorInput::Json { url: src.parse()?, data: content })
                    }
                    ImportSource::Proto => {
                        let metadata = proto_reader.read(&src).await?;
                        generator_type_inputs.push(GeneratorInput::Proto { metadata });
                    }
                }
            }
            InputSource::Config { src, _marker } => {
                let source = config::Source::detect(&src)?;
                let schema = reader.read_file(&src).await?.content;
                generator_type_inputs.push(GeneratorInput::Config { schema, source });
            }
        }
    }

    Ok(generator_type_inputs)
}

pub async fn run_test(path: &str) -> anyhow::Result<()> {
    let contents = std::fs::read_to_string(path)?;
    let config: GeneratorTest = serde_json::from_str(&contents)?;

    let resolved_config = config.config.resolve_paths(path)?;
    let generator_inp = resolve_io(resolved_config, config.resolvers).await?;
    let gen = GeneratorReImpl::new("f", "T");
    let cfg_module = gen.run("Query", &generator_inp)?;

    insta::assert_snapshot!(path, cfg_module.config.to_sdl());

    Ok(())
}
