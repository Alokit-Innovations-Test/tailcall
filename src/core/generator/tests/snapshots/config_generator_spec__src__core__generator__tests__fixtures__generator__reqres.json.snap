---
source: src/core/generator/tests/config_generator_spec.rs
expression: config.to_sdl()
---
schema @server @upstream(baseURL: "https://reqres.in") {
  query: Query
}

type Daum {
  avatar: String
  email: String
  first_name: String
  id: Int
  last_name: String
}

type Query {
  user(p1: Int!): T6 @http(path: "/api/users/{{.args.p1}}")
  users(page: Int): User @http(path: "/api/users", query: [{key: "page", value: "{{.args.page}}"}])
}

type Support {
  text: String
  url: String
}

type T6 {
  data: Daum
  support: Support
}

type User {
  data: [Daum]
  page: Int
  per_page: Int
  support: Support
  total: Int
  total_pages: Int
}
